numpy/lib/function_base.py

```diff
@@ -689,16 +699,38 @@ def histogram(a, bins=10, range=None, normed=False, weights=None,
                 b = a[keep]
 
         if b.size == 0:
-            bins = 1
+            n_equal_bins = 1
         else:
             # Do not call selectors on empty arrays
-            width = _hist_bin_selectors[bins](b)
+            width = _hist_bin_selectors[bin_name](b)
             if width:
-                bins = int(np.ceil((mx - mn) / width))
+                n_equal_bins = int(np.ceil((mx - mn) / width))
             else:
                 # Width can be zero for some estimators, e.g. FD when
                 # the IQR of the data is zero.
-                bins = 1
+                n_equal_bins = 1
+
+    elif np.ndim(bins) == 0:
+        try:
+            n_equal_bins = operator.index(bins)
+        except TypeError:
+            raise TypeError('If a scalar, `bins` must be an integer')
+        if n_equal_bins < 1:
+            raise ValueError('If an integer, `bins` must be positive')
+
+    elif np.ndim(bins) == 1:
+        bin_edges = np.asarray(bins)
+        if np.any(bin_edges[:-1] > bin_edges[1:]):
+            raise ValueError('`bins` must increase monotonically')
+
+    else:
+        raise ValueError("If an array, `bins` must be 1d")
+
+    del bins
+
+    # compute the bins if only the count was specified
+    if n_equal_bins is not None:
+        bin_edges = linspace(mn, mx, n_equal_bins + 1, endpoint=True)
```

Comment:
Maybe `binmin, binmax` instead of `mn, mx`?

Reply:
Not all they're used for though

Reply:
`(first|last) _edge` sounds good.

Reply:
Done